sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,769 @@
+            if n_comps[i] > X.shape[0]:
+                msg = names[i] + "must be <= n_samples, but" + names[i]
+                msg += "= {}, n_samples = {}".format(n_comps[i], X.shape[0])
+                raise ValueError(msg)
+
+        # check if X contains the 0 vector
+        if np.any(~X.any(axis=1)) and (
+            "cosine" in self.affinity or self.affinity == "all"
+        ):
+            if isinstance(self.affinity, np.ndarray):
+                self.affinity = np.delete(
+                    self.affinity, np.argwhere(self.affinity == "cosine")
+                )
+            if isinstance(self.affinity, list):
+                self.affinity.remove("cosine")
+            if self.affinity == "all":
+                self.affinity = ["euclidean", "manhattan", "none"]
+            warnings.warn(
+                "X contains a zero vector, will not run cosine affinity."
+            )
+
+        label_init = self.label_init
+        if label_init is not None:
+            if label_init.size != X.shape[0]:
+                msg = "n_samples must be the same as the length of label_init"
+                raise ValueError(msg)
+
+        param_grid = dict(
+            n_components=range(self.min_components, self.max_components + 1),
+            random_state=[self.random_state],
+        )
+        param_grid["affinity"] = (
+            self.affinity
+            if self.affinity != "all"
+            else ["euclidean", "manhattan", "cosine", "none"]
+        )
+        param_grid["linkage"] = (
+            self.linkage
+            if self.linkage != "all"
+            else ["ward", "complete", "average", "single"]
+        )
+        param_grid["covariance_type"] = (
+            self.covariance_type
+            if self.covariance_type != "all"
+            else ["spherical", "diag", "tied", "full"]
+        )
+
+        param_grid = list(ParameterGrid(param_grid))
+        param_grid_ag, param_grid = _process_paramgrid(param_grid)
+
+        n = X.shape[0]
+        if self.max_agglom_size is None or n <= self.max_agglom_size:
+            X_subset = X
+        else:  # if dataset is huge, agglomerate a subset
+            subset_idxs = np.random.choice(
+                np.arange(0, n), self.max_agglom_size
+            )
+            X_subset = X[subset_idxs, :]
+
+        ag_labels = []
+        if self.label_init is None:
+            for p_ag in param_grid_ag:
+                if p_ag["affinity"] != "none":
+                    agg = AgglomerativeClustering(
+                        n_clusters=self.min_components, **p_ag
+                    )
+                    agg.fit(X_subset)
+                    hierarchical_labels = _hierarchical_labels(
+                        agg.children_, self.min_components, self.max_components
+                    )
+                    ag_labels.append(hierarchical_labels)
+
+        def _fit_for_data(p):
+            ag_params = p[0]
+            gm_params = p[1]
```

Comment:
This relates to [my comment](https://github.com/scikit-learn/scikit-learn/pull/19562/files#r584360251).